query-engine/core/src/interactive_transactions/manager.rs

```diff
@@ -111,24 +111,33 @@ impl ItxManager {
         isolation_level: Option<String>,
         timeout: Duration,
     ) -> crate::Result<()> {
-        // This task notifies the task spawned in `new()` method that the timeout for this
-        // transaction has expired.
-        crosstarget_utils::task::spawn({
-            let timeout_sender = self.timeout_sender.clone();
-            let tx_id = tx_id.clone();
-            async move {
-                crosstarget_utils::time::sleep(timeout).await;
-                timeout_sender.send(tx_id).expect("receiver must exist");
-            }
-        });
+        // Only create a client if there is no client for this transaction yet.
+        // otherwise, begin a new transaction/savepoint for the existing client.
+        if self.transactions.read().await.contains_key(&tx_id) {
```

Comment:
This will hold the lock for the whole duration of the `if` statement (and, since, we haven't migrated this crate to Rust 2024, even the `else` branch if the condition is false). You should extract the condition into a variable so that the temporary is dropped at the end of the assignment statement.

Reply:
I'm actually not sure if/why this works right now: it looks like it should deadlock in the `else` branch when trying to do `self.transactions.write()`.